File: experiment-10/brazil/runs/language=go_model=claude-fable-5/rep1/store_test.go
// store_test.go — BDD (Given/When/Then) scenarios for data loading,
// deduplication and team-name normalization.
package main

import (
	"sync"
	"testing"
)

var (
	testStoreOnce sync.Once
	testStore     *Store
	testStoreErr  error
)

// givenTheDataIsLoaded loads the CSV datasets once and shares the store
// across all test scenarios.
func givenTheDataIsLoaded(t *testing.T) *Store {
	t.Helper()
	testStoreOnce.Do(func() {
		testStore, testStoreErr = LoadStore("data/kaggle")
	})
	if testStoreErr != nil {
		t.Fatalf("Given the match data is loaded: %v", testStoreErr)
	}
	return testStore
}

func TestScenario_AllSixDatasetsAreLoadable(t *testing.T) {
	// Given the six Kaggle CSV files in data/kaggle
	// When the store is loaded
	s := givenTheDataIsLoaded(t)

	// Then matches from every match file are present
	bySource := map[string]int{}
	for _, m := range s.Matches {
		bySource[m.Source]++
	}
	for _, src := range []string{
		"Brasileirao_Matches.csv",
		"novo_campeonato_brasileiro.csv",
		"Brazilian_Cup_Matches.csv",
		"Libertadores_Matches.csv",
		"BR-Football-Dataset.csv",
	} {
		if bySource[src] == 0 {
			t.Errorf("expected matches loaded from %s, got none", src)
		}
	}
	// And the FIFA player file is loaded
	if len(s.Players) < 18000 {
		t.Errorf("expected ~18207 players, got %d", len(s.Players))
	}
	if len(s.Matches) < 10000 {
		t.Errorf("expected >10000 deduplicated matches, got %d", len(s.Matches))
	}
	t.Logf("loaded %d matches (%v) and %d players", len(s.Matches), bySource, len(s.Players))
}

func TestScenario_TeamNameVariationsNormalizeToTheSameKey(t *testing.T) {
	// Given team names as spelled in the different datasets
	// When each is normalized
	// Then variations of the same club share one key
	same := [][]string{
		{"Palmeiras-SP", "Palmeiras", "palmeiras"},
		{"Flamengo-RJ", "Flamengo", "FLAMENGO"},
		{"Sao Paulo-SP", "São Paulo", "Sao Paulo"},
		{"Gremio-RS", "Grêmio", "Gremio"},
		{"Vasco da Gama-RJ", "Vasco", "vasco da gama"},
		{"Athletico Paranaense - PR", "Atletico-PR", "Athletico-PR", "Athletico Paranaense"},
		{"Atletico-MG", "Atlético-MG", "Atletico Mineiro"},
		{"Avai-SC", "Avaí"},
		{"Ceara-CE", "Ceará"},
		{"Red Bull Bragantino-SP", "Bragantino"},
		{"America - MG", "América-MG", "America MG"},
	}
	for _, group := range same {
		want := teamKey(group[0])
		for _, name := range group[1:] {
			if got := teamKey(name); got != want {
				t.Errorf("teamKey(%q) = %q, want %q (same as %q)", name, got, want, group[0])
			}
		}
	}
}

func TestScenario_AmbiguousClubsStayDistinct(t *testing.T) {
	// Given clubs that share a base name but are different teams
	// When their names are normalized
	// Then they do NOT collide
	distinct := [][2]string{
		{"Atlético-MG", "Atlético-GO"},
		{"Atlético-MG", "Athletico-PR"},
		{"América-MG", "América-RN"},
		{"Botafogo-RJ", "Botafogo PB"},
		{"Botafogo-RJ", "Botafogo SP"},
		{"Grêmio", "Grêmio Prudente"},
	}
	for _, pair := range distinct {
		if k1, k2 := teamKey(pair[0]), teamKey(pair[1]); k1 == k2 {
			t.Errorf("teamKey(%q) == teamKey(%q) == %q; want distinct keys", pair[0], pair[1], k1)
		}
	}
}

func TestScenario_UTF8AccentsAreHandled(t *testing.T) {
	// Given Brazilian Portuguese names with accents and cedillas
	cases := map[string]string{
		"São Paulo":               "sao paulo",
		"Grêmio":                  "gremio",
		"Avaí":                    "avai",
		"Fortaleza Esporte Clube": "fortaleza esporte clube",
		"Criciúma":                "criciuma",
		"Goiás":                   "goias",
	}
	// When normalized / Then accents are transliterated
	for in, want := range cases {
		if got := normalizeText(in); got != want {
			t.Errorf("normalizeText(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestScenario_MultipleDateFormatsAreParsed(t *testing.T) {
	// Given the three date formats used across the datasets
	cases := map[string]string{
		"2023-09-24":          "2023-09-24",
		"29/03/2003":          "2003-03-29",
		"2012-05-19 18:30:00": "2012-05-19",
	}
	// When parsed / Then all succeed
	for in, want := range cases {
		d, ok := parseDate(in)
		if !ok {
			t.Errorf("parseDate(%q) failed", in)
			continue
		}
		if got := d.Format("2006-01-02"); got != want {
			t.Errorf("parseDate(%q) = %s, want %s", in, got, want)
		}
	}
}

func TestScenario_OverlappingSerieASeasonsAreDeduplicated(t *testing.T) {
	// Given Série A 2012-2019 appears in Brasileirao_Matches.csv,
	// novo_campeonato_brasileiro.csv and BR-Football-Dataset.csv
	s := givenTheDataIsLoaded(t)

	// When counting Série A matches per overlapping season
	// Then each season has at most 380 matches (20 teams, double round-robin)
	for season := 2012; season <= 2019; season++ {
		n := 0
		for _, m := range s.Matches {
			if m.Competition == CompSerieA && m.Season == season {
				n++
			}
		}
		if n > 380 {
			t.Errorf("season %d: %d Série A matches after dedup, want <= 380", season, n)
		}
		if n < 370 {
			t.Errorf("season %d: only %d Série A matches, expected a full season (~380)", season, n)
		}
	}
}

func TestScenario_ExtendedStatsAreMergedIntoDeduplicatedMatches(t *testing.T) {
	// Given BR-Football-Dataset.csv carries shots/corners for matches that
	// also exist in other files
	s := givenTheDataIsLoaded(t)

	// Then at least some stored matches expose extended stats
	n := 0
	for _, m := range s.Matches {
		if m.HasStats() {
			n++
		}
	}
	if n == 0 {
		t.Error("expected some matches with extended stats (shots/corners), got none")
	}
}

File: experiment-15-sonnet5/brazil/runs/language=typescript_model=sonnet-5_prompt=bdd/rep1/tests/store.test.ts
import { beforeAll, describe, expect, it } from "vitest";
import type { SoccerDataStore } from "../src/data/store.js";
import { loadTestStore } from "./support/testStore.js";

describe("SoccerDataStore", () => {
  let store: SoccerDataStore;

  beforeAll(async () => {
    store = await loadTestStore();
  });

  it("test_given_all_six_source_files_when_loaded_then_matches_and_players_are_present", () => {
    // Given the six provided Kaggle CSV files
    // When the store loads them all
    // Then both match and player data are non-empty
    expect(store.matches.length).toBeGreaterThan(0);
    expect(store.players.length).toBeGreaterThan(0);
  });

  it("test_given_all_fifa_rows_when_loaded_then_every_player_row_is_kept", () => {
    // Given fifa_data.csv has 18,207 player rows
    // When the store loads the file
    // Then every player row is present (no accidental drops)
    expect(store.players.length).toBe(18207);
  });

  it("test_given_overlapping_brasileirao_sources_when_loaded_then_a_season_is_not_double_counted", () => {
    // Given Brasileirao_Matches.csv and novo_campeonato_brasileiro.csv both independently
    // cover full 2012-2019 seasons (380 matches per season each)
    // When the store loads both files
    const matches2019 = store.matches.filter((m) => m.competition === "Brasileirao" && m.season === 2019);
    // Then the season appears once (380 matches), not twice (760)
    expect(matches2019.length).toBe(380);
  });

  it("test_given_overlapping_copa_do_brasil_sources_when_loaded_then_a_season_is_not_double_counted", () => {
    // Given Brazilian_Cup_Matches.csv and BR-Football-Dataset.csv both cover 2016 Copa do Brasil matches
    // When the store loads both files
    const matches2016 = store.matches.filter((m) => m.competition === "CopaDoBrasil" && m.season === 2016);
    // Then the primary source's count (162) is used rather than a doubled total
    expect(matches2016.length).toBe(162);
  });

  it("test_given_a_club_present_in_both_match_and_player_data_when_looked_up_by_key_then_both_link_to_the_same_team_key", () => {
    // Given Grêmio appears in both match data (as "Gremio" without accents) and FIFA player data (as "Grêmio")
    const matches = store.matchesForTeamKey("gremio");
    const players = store.playersForClubKey("gremio");
    // When looking both up by the shared normalized key
    // Then both non-trivial result sets are found under the same key
    expect(matches.length).toBeGreaterThan(0);
    expect(players.length).toBeGreaterThan(0);
  });

  it("test_given_a_row_with_an_unparseable_date_when_loaded_then_the_match_is_kept_with_a_null_date", () => {
    // Given Libertadores_Matches.csv contains at least one row with datetime "NA"
    // When the store loads it
    const unparsed = store.matches.filter((m) => m.date === null);
    // Then the row is retained (not silently dropped) with a null date rather than throwing
    expect(unparsed.length).toBeGreaterThan(0);
  });
});
